File: packages/typescript/tests/capture.test.ts
// Why: capture() is the central producer entry point. Its outputs feed every
// downstream consumer; these tests pin the chain semantics from
// `tests/conformance/features/chain.feature` plus the JS-specific quirks
// (`Error.cause`, `AggregateError.errors`) the Rust reference can't exercise.
import { describe, it, expect } from 'vitest';
import { capture } from '../src/capture.js';

describe('capture: chain semantics', () => {
  it('single error produces a one-entry chain with exception_id=0, parent_id=0', () => {
    // Why: chain.feature "Single error produces a one-entry chain"
    const chain = capture(new Error('boom'), { release: 'r', serverName: 's' });
    expect(chain).toHaveLength(1);
    const mech = chain[0]!.mechanism!;
    expect(mech.exception_id).toBe(0);
    expect(mech.parent_id).toBe(0);
  });

  it('three-deep chain via Error.cause has correct mechanism IDs', () => {
    // Why: chain.feature "Three-deep chain has correct mechanism IDs" —
    // exception_id is positional, parent_id stamps the causal link.
    const inner = new Error('inner');
    const middle = new Error('middle', { cause: inner });
    const outer = new Error('outer', { cause: middle });
    const chain = capture(outer, { release: 'r', serverName: 's' });
    expect(chain).toHaveLength(3);
    expect(chain[0]!.mechanism!.exception_id).toBe(0);
    expect(chain[0]!.mechanism!.parent_id).toBe(0);
    expect(chain[1]!.mechanism!.exception_id).toBe(1);
    expect(chain[1]!.mechanism!.parent_id).toBe(0);
    expect(chain[2]!.mechanism!.exception_id).toBe(2);
    expect(chain[2]!.mechanism!.parent_id).toBe(1);
    // most-causal-first → innermost is index 0, outermost is last
    expect(chain[0]!.message).toBe('inner');
    expect(chain[2]!.message).toBe('outer');
  });

  it('last chain entry is the originating caught (outermost) error', () => {
    // Why: chain.feature "Originating caught error is the last element" —
    // mirrors Sentry's `exception.values` convention.
    const inner = new Error('inner');
    const outer = new Error('outer', { cause: inner });
    const chain = capture(outer, { release: 'r', serverName: 's' });
    expect(chain[chain.length - 1]!.message).toBe('outer');
  });

  it('default mechanism is generic + handled=true', () => {
    // Why: matches Rust reference impl (sererr/src/lib.rs `capture()`).
    const chain = capture(new Error('x'), { release: 'r', serverName: 's' });
    expect(chain[0]!.mechanism!.type).toBe('generic');
    expect(chain[0]!.mechanism!.handled).toBe(true);
  });

  it('leaf type falls back to err.constructor.name', () => {
    // Why: with no typeName option, use the runtime class name (TypeError, etc.)
    const chain = capture(new TypeError('bad'), { release: 'r', serverName: 's' });
    expect(chain[chain.length - 1]!.type).toBe('TypeError');
  });

  it('leaf typeName option overrides err.constructor.name', () => {
    // Why: producers at the catch site may know a more specific type
    // (matches Rust signature `capture(&err, type_name, ...)`).
    const chain = capture(new Error('x'), {
      typeName: 'MyDomainError',
      release: 'r',
      serverName: 's',
    });
    expect(chain[chain.length - 1]!.type).toBe('MyDomainError');
  });

  it('release and serverName are stamped on every chain entry', () => {
    // Why: every CapturedError is self-contained per the proto spec
    // (release / server_name are inlined event-level fields).
    const inner = new Error('inner');
    const outer = new Error('outer', { cause: inner });
    const chain = capture(outer, { release: 'v1.2.3', serverName: 'pod-7' });
    for (const entry of chain) {
      expect(entry.release).toBe('v1.2.3');
      expect(entry.serverName ?? entry.server_name).toBeDefined();
      expect(entry.server_name).toBe('pod-7');
    }
  });

  it('non-Error throwables capture as a single entry', () => {
    // Why: `throw "string"` and `throw 42` are legal in JS; capture must
    // not crash and must produce a one-entry chain with a sensible message.
    const chain = capture('plain string failure', { release: 'r', serverName: 's' });
    expect(chain).toHaveLength(1);
    expect(chain[0]!.message).toBe('plain string failure');
  });

  it('AggregateError flattens errors as siblings sharing parent_id', () => {
    // Why: AggregateError carries plural `.errors` from Promise.any / allSettled.
    // Idiosyncrasy noted in the task spec — siblings must share the outermost's
    // parent_id so consumers can reconstruct the tree.
    const e1 = new Error('first');
    const e2 = new Error('second');
    const agg = new AggregateError([e1, e2], 'all failed');
    const chain = capture(agg, { release: 'r', serverName: 's' });
    // Outermost is the AggregateError; siblings are flattened in.
    expect(chain.length).toBeGreaterThanOrEqual(3);
    // The AggregateError itself is the outermost (last).
    expect(chain[chain.length - 1]!.message).toBe('all failed');
    // The siblings share the same parent_id (the AggregateError's exception_id).
    const aggId = chain[chain.length - 1]!.mechanism!.exception_id;
    const sibs = chain.filter((c) => c.message === 'first' || c.message === 'second');
    expect(sibs.length).toBe(2);
    for (const s of sibs) {
      expect(s.mechanism!.parent_id).toBe(aggId);
    }
  });
});

describe('capture: frames', () => {
  it('captures at least one frame from a thrown Error', () => {
    // Why: chain.feature "Frame ordering is most-recent-first" — at minimum,
    // assert frames exist (cross-language contract; specific function names
    // are runtime-dependent).
    let caught: Error | undefined;
    function deep(): void {
      throw new Error('boom');
    }
    try {
      deep();
    } catch (e) {
      caught = e as Error;
    }
    const chain = capture(caught!, { release: 'r', serverName: 's' });
    expect(chain[0]!.frames.length).toBeGreaterThan(0);
  });

  it('first frame is the most recent call', () => {
    // Why: chain.feature pins frame ordering. V8 stack format already lists
    // the throw site first; we must not reverse it.
    let caught: Error | undefined;
    function inner(): void {
      throw new Error('boom');
    }
    function outer(): void {
      inner();
    }
    try {
      outer();
    } catch (e) {
      caught = e as Error;
    }
    const chain = capture(caught!, { release: 'r', serverName: 's' });
    const frames = chain[0]!.frames;
    // The throwing function (inner) should appear before its caller (outer).
    const innerIdx = frames.findIndex((f) => f.function.includes('inner'));
    const outerIdx = frames.findIndex((f) => f.function.includes('outer'));
    expect(innerIdx).toBeGreaterThanOrEqual(0);
    expect(outerIdx).toBeGreaterThanOrEqual(0);
    expect(innerIdx).toBeLessThan(outerIdx);
  });
});

File: packages/csharp/tests/CaptureTest.cs
// Pins the Capture.Build semantics:
//  - single exception → one chain entry
//  - InnerException walked → most-causal-first array
//  - mechanism IDs stamped (exception_id, parent_id)
//  - default mechanism (type="generic", handled=true)
//  - AggregateException flattening
//  - frames captured from the exception's stack trace,
//    most-recent-call-first.

using FluentAssertions;
using Sererr;
using Xunit;

namespace Sererr.Tests;

public class CaptureTest
{
    private static Exception ThrowMe(string msg)
    {
        try { throw new InvalidOperationException(msg); }
        catch (Exception e) { return e; }
    }

    [Fact]
    // A single thrown exception captures to a one-entry chain whose
    // mechanism.exception_id == 0 and parent_id == 0.
    public void SingleException_ProducesOneEntryChain()
    {
        var ex = ThrowMe("oops");
        var chain = Capture.Build(ex, "InvalidOperationException", "v1", "host");

        chain.Should().HaveCount(1);
        chain[0].Type.Should().Be("InvalidOperationException");
        chain[0].Message.Should().Be("oops");
        chain[0].Mechanism.Should().NotBeNull();
        chain[0].Mechanism!.ExceptionId.Should().Be(0u);
        chain[0].Mechanism!.ParentId.Should().Be(0u);
    }

    [Fact]
    // Default mechanism: Type="generic", Handled=true.
    public void DefaultMechanism_IsGenericHandled()
    {
        var ex = ThrowMe("x");
        var chain = Capture.Build(ex, "T", "r", "s");
        chain[0].Mechanism!.Type.Should().Be("generic");
        chain[0].Mechanism!.Handled.Should().BeTrue();
    }

    [Fact]
    // Three-deep InnerException chain → 3-entry chain, most-causal-first
    // (innermost is element 0). Mechanism IDs stamped i / max(0,i-1).
    public void InnerExceptionChain_FlattenedMostCausalFirst()
    {
        Exception inner;
        try { throw new InvalidOperationException("inner"); }
        catch (Exception e) { inner = e; }

        Exception middle;
        try { throw new InvalidOperationException("middle", inner); }
        catch (Exception e) { middle = e; }

        Exception outer;
        try { throw new InvalidOperationException("outer", middle); }
        catch (Exception e) { outer = e; }

        var chain = Capture.Build(outer, "OuterError", "r", "s");

        chain.Should().HaveCount(3);
        chain[0].Message.Should().Be("inner");
        chain[2].Message.Should().Be("outer");
        chain[0].Mechanism!.ExceptionId.Should().Be(0u);
        chain[0].Mechanism!.ParentId.Should().Be(0u);
        chain[1].Mechanism!.ExceptionId.Should().Be(1u);
        chain[1].Mechanism!.ParentId.Should().Be(0u);
        chain[2].Mechanism!.ExceptionId.Should().Be(2u);
        chain[2].Mechanism!.ParentId.Should().Be(1u);
    }

    [Fact]
    // The leaf chain entry (the originating caught error) gets the
    // user-supplied type_name; other entries get GetType().FullName.
    public void TypeName_AppliedToLeaf()
    {
        Exception inner;
        try { throw new InvalidOperationException("inner"); }
        catch (Exception e) { inner = e; }

        Exception outer;
        try { throw new InvalidOperationException("outer", inner); }
        catch (Exception e) { outer = e; }

        var chain = Capture.Build(outer, "MyCustomLabel", "r", "s");

        // The outermost (caught) error is last; that's the one carrying
        // the operator-supplied label.
        chain[^1].Type.Should().Be("MyCustomLabel");
        chain[0].Type.Should().Be(typeof(InvalidOperationException).FullName);
    }

    [Fact]
    // AggregateException.InnerExceptions is plural: each inner is a
    // sibling sharing the AggregateException as parent. Flattened in
    // order, all addressing the AggregateException's parent_id.
    public void AggregateException_FlattensSiblings()
    {
        Exception a;
        try { throw new InvalidOperationException("a"); }
        catch (Exception e) { a = e; }
        Exception b;
        try { throw new InvalidOperationException("b"); }
        catch (Exception e) { b = e; }

        Exception agg;
        try { throw new AggregateException("agg", a, b); }
        catch (Exception e) { agg = e; }

        var chain = Capture.Build(agg, "AggregateException", "r", "s");

        // 3 entries: a, b (siblings, most-causal-first), then agg.
        chain.Should().HaveCount(3);
        // The two siblings come first; aggregate is last.
        chain[^1].Message.Should().StartWith("agg");
        // Both siblings parent the aggregate (which sits at index 2).
        chain[0].Mechanism!.ParentId.Should().Be(2u);
        chain[1].Mechanism!.ParentId.Should().Be(2u);
        chain[2].Mechanism!.ParentId.Should().Be(0u);
        chain[2].Mechanism!.ExceptionId.Should().Be(2u);
    }

    [Fact]
    // Frames are captured most-recent-first; the throwing frame is at
    // index 0. We assert the first frame's function name contains the
    // helper that threw.
    public void Frames_MostRecentFirst()
    {
        var ex = ThrowMe("x");
        var chain = Capture.Build(ex, "T", "r", "s");
        var frames = chain[0].Frames;
        frames.Should().NotBeEmpty();
        // Top frame is the throwing helper.
        frames[0].Function.Should().Contain("ThrowMe");
    }

    [Fact]
    // Release / server_name flow through to every entry in the chain.
    public void ReleaseAndServerName_AppliedToEveryEntry()
    {
        Exception inner;
        try { throw new InvalidOperationException("inner"); }
        catch (Exception e) { inner = e; }
        Exception outer;
        try { throw new InvalidOperationException("outer", inner); }
        catch (Exception e) { outer = e; }

        var chain = Capture.Build(outer, "T", "v1.0.0", "host-a");
        chain.Should().AllSatisfy(c =>
        {
            c.Release.Should().Be("v1.0.0");
            c.ServerName.Should().Be("host-a");
        });
    }
}
